This patch provides a way to get the exception object on a "raise"
event in a trace hook. The exeception object passed as the "klass"
parameter in the hook callback.

Note that in a hook one can get the class via a binding from the frame
using eval('self.class', binding)

Another small but unrelated change here is to fill in some of the
missing event names in the case statement of routine get_event_name().


Index: eval.c
===================================================================
--- eval.c	(revision 28955)
+++ eval.c	(working copy)
@@ -440,7 +440,12 @@
     rb_trap_restore_mask();
 
     if (tag != TAG_FATAL) {
-	EXEC_EVENT_HOOK(th, RUBY_EVENT_RAISE, th->cfp->self, 0, 0);
+        /* In a RAISE event, we store as the "class" parameter the
+           the optional message parameter which is likely to be of more
+           use. Given a binding, a trace hook can get the class via 
+           eval('self.class', binding)
+	 */
+        EXEC_EVENT_HOOK(th, RUBY_EVENT_RAISE, th->cfp->self, 0, mesg);
     }
 }
 
Index: thread.c
--- thread.c	(revision 28955)
+++ thread.c	(working copy)
===================================================================
@@ -4046,6 +4109,12 @@
 	return "c-return";
       case RUBY_EVENT_RAISE:
 	return "raise";
+      case RUBY_EVENT_SWITCH:
+        return "switch";
+      case RUBY_EVENT_COVERAGE:
+        return "coverage";
+      case RUBY_EVENT_VM:
+        return "vm";
       default:
 	return "unknown";
     }
@@ -4077,8 +4055,15 @@
 	p->event == RUBY_EVENT_C_RETURN) {
 	id = p->id;
 	klass = p->klass;
-    }
-    else {
+    } else if (p->event == RUBY_EVENT_RAISE) {
+        /* rb_thread_method_and_id() wants a place to store a klass
+	    value which subsequently we will not use.  */
+        VALUE junk_klass; 
+	 rb_thread_method_id_and_class(GET_THREAD(), &id, &junk_klass);
+	 /* When the event is RUBY_EVENT_RAISE, we have stored the raise
+	    message as the trace-hook klass parameter. */ 
+	 klass = p->klass;
+    } else {
 	rb_thread_method_id_and_class(GET_THREAD(), &id, &klass);
     }
     if (id == ID_ALLOCATOR)
Index: test/ruby/test_settracefunc.rb
===================================================================
--- test/ruby/test_settracefunc.rb	(revision 28955)
+++ test/ruby/test_settracefunc.rb	(working copy)
@@ -220,7 +220,7 @@
      3: })
      4: begin
      5:   raise TypeError, "error"
-     6: rescue TypeError
+     6: rescue TypeError => $e
      7: end
      8: set_trace_func(nil)
     EOF
@@ -248,7 +248,7 @@
                  events.shift)
     assert_equal(["c-return", 5, :set_backtrace, Exception],
                  events.shift)
-    assert_equal(["raise", 5, :test_raise, TestSetTraceFunc],
+    assert_equal(["raise", 5, :test_raise, $e], 
                  events.shift)
     assert_equal(["c-return", 5, :raise, Kernel],
                  events.shift)

Index: test/ruby/test_tracefunc_raise.rb
===================================================================
--- test/ruby/test_tracefunc_raise.rb	(revision 28955)
+++ test/ruby/test_tracefunc_raise.rb	(working copy)
@@ -0,0 +1,26 @@
+require 'test/unit'
+
+# tests that we a trace hook has access to the runtime exception Object
+# when it is called through a raise event
+
+class TestTracefuncRaise < Test::Unit::TestCase
+
+  def test_basic
+    tuples = []
+    p = Proc.new { 
+      |event, file, lineno, mid, binding, klass|
+      tuples << klass
+    }
+    msg = 'this is a message'
+    set_trace_func(p, 0x0080)
+    begin ; x = 1/0; rescue; end
+    begin ; raise RuntimeError, msg; rescue; end
+    clear_trace_func
+    assert_equal(2, tuples.size, 
+                 "Wrong number of tuples captured #{tuples.inspect}")
+    assert_equal msg, tuples[1].message
+    assert_equal([ZeroDivisionError, RuntimeError], tuples.map{|t| t.class},
+                 "Mismatched tuples classes in #{tuples.inspect}")
+    
+  end
+end
